src/nupic/research/temporal_memory.py

```diff
@@ -120,419 +209,192 @@ def compute(self, activeColumns, learn=True):
-
-    # Sort unpredictedActiveColumns before iterating for compatibility with C++
-    for column in sorted(unpredictedActiveColumns):
-      cells = self.cellsForColumn(column)
-      activeCells.update(cells)
-
-      (bestCell,
-       bestSegment) = self.bestMatchingCell(cells,
-                                            prevActiveCells,
-                                            connections)
-      winnerCells.add(bestCell)
-
-      if bestSegment is None and len(prevWinnerCells):
-        bestSegment = connections.createSegment(bestCell)
-
-      if bestSegment is not None:
-        learningSegments.add(bestSegment)
-
-    return activeCells, winnerCells, learningSegments
-
-
-  def learnOnSegments(self,
-                      prevActiveSegments,
-                      learningSegments,
-                      prevActiveCells,
-                      winnerCells,
-                      prevWinnerCells,
-                      connections,
-                      predictedInactiveCells,
-                      prevMatchingSegments):
-    """
-    Phase 3: Perform learning by adapting segments.
-
-    Pseudocode:
-
-      - (learning) for each prev active or learning segment
-        - if learning segment or from winner cell
-          - strengthen active synapses
-          - weaken inactive synapses
-        - if learning segment
-          - add some synapses to the segment
-            - subsample from prev winner cells
-
-      - if predictedSegmentDecrement > 0
-        - for each previously matching segment
-          - if cell is a predicted inactive cell
-            - weaken active synapses but don't touch inactive synapses
-
-    @param prevActiveSegments           (set)         Indices of active segments in `t-1`
-    @param learningSegments             (set)         Indices of learning segments in `t`
-    @param prevActiveCells              (set)         Indices of active cells in `t-1`
-    @param winnerCells                  (set)         Indices of winner cells in `t`
-    @param prevWinnerCells              (set)         Indices of winner cells in `t-1`
-    @param connections                  (Connections) Connectivity of layer
-    @param predictedInactiveCells       (set)         Indices of predicted inactive cells
-    @param prevMatchingSegments         (set)         Indices of segments with
+                      `cells`         (list),
+                      `bestCell`      (int),
     """
-    segments = prevActiveSegments | learningSegments
-
-    # Sort segments before iterating for compatibility with C++
-    # Sort with primary key = cell idx, secondary key = segment idx
-    segments = sorted(
-      segments,
-      key=lambda segment: (connections.cellForSegment(segment), segment))
-
-    for segment in segments:
-      isLearningSegment = segment in learningSegments
-      isFromWinnerCell = connections.cellForSegment(segment) in winnerCells
+    cells = self.cellsForColumn(excitedColumn["column"])
```

Comment:
This line can be moved to the `else`. That will help perf. Really we could just pass the column number into `leastUsedCell`.

Also, `cellsForColumn` is slow since it doesn't use an xrange, but that's independent of this change. I'd like to see us stop using `cellsForColumn` as part of the algorithm, but we don't have to do that now.